cs_tests/cs15.c

```diff
@@ -22,6 +19,8 @@ int main(){
 	c = &b1;
 	d = &d1;
 	bar(a,b,c,d);
+	MUSTALIAS(a1,&d1);
```

Comment:
As far as I understand, a1 must points to d1, because of the statement `*w = t` in the function _foo_.
And since the pointers are being updated in foo and hence they should also be reflected in main.
Hence the MUSTALIAS(a1,&d1).

Please correct me if I am wrong.

Reply:
If you change `d = &d1;` to be `b1=&d1` then it looks correct. A simple way to test the correctness is to print the values of `a1` and `&d1` at runtime (e..g, using `printf("%d,%d", a, &d1)`, you will know.

Reply:
Hi,
Thanks for the suggestion! I did as you suggested, and came up with the following output for this test case.
Before function call bar() :
```
Value of d1 = 21, *a1 = 29590344, *b1 = -443987883
Address of d1 = -398197836, *a1 = 1595504733, *b1 = -2012571232
```
After function call bar() :
```
Value of d1 = 21, *a1 = 21, *b1 = 21
Address of d1 = -398197836, *a1 = -398197836, *b1 = -398197836
```
It can be seen that the alias functions (`MUSTALIAS(a1,&d1), MUSTALIAS(b1,a1)`) are correct.



Also sharing the code for main():
```
int main(){

        int **a,**b,**c,*d,*a1,*b1,*c1,d1;
        d1=21;
        printf("Value of d1 = %d, *a1 = %d, *b1 = %d\n",d1,*a1,*b1);
        printf("Address of d1 = %d, *a1 = %d, *b1 = %d\n",&d1,a1,b1);
        a = &a1;
        b = &b1;
        c = &b1;
        d = &d1;
        bar(a,b,c,d);
        printf("Value of d1 = %d, *a1 = %d, *b1 = %d\n",d1,*a1,*b1);
        printf("Address of d1 = %d, *a1 = %d, *b1 = %d\n",&d1,a1,b1);
        return 0;
        //MUSTALIAS(a1,&d1);
        //MUSTALIAS(b1,a1);
//      foo(a,b,c,d);
//      foo(b,a,c,d);
}
```

Reply:
The code is incomplete. What is the code in `bar` and `foo`?

Reply:
I have not made any changes to the functions` bar` and `foo`, just made changes to `main` and hence shared the code for `main`. 
`foo` and `bar` are the same as they were in cs15.c

Reply:
Sharing the whole code :
```
#include <stdio.h>
void foo(int **w,int**x,int **y, int *z){
	int *t;
	*y = z;
	t = *x;
	*w = t;

}

void bar(int **p1,int**p2,int **p3, int *p4){
	foo(p1,p2,p3,p4);
}
int main(){

        int **a,**b,**c,*d,*a1,*b1,*c1,d1;
        d1=21;
        printf("Value of d1 = %d, *a1 = %d, *b1 = %d\n",d1,*a1,*b1);
        printf("Address of d1 = %d, *a1 = %d, *b1 = %d\n",&d1,a1,b1);
        a = &a1;
        b = &b1;
        c = &b1;
        d = &d1;
        bar(a,b,c,d);
        printf("Value of d1 = %d, *a1 = %d, *b1 = %d\n",d1,*a1,*b1);
        printf("Address of d1 = %d, *a1 = %d, *b1 = %d\n",&d1,a1,b1);
        return 0;
        //MUSTALIAS(a1,&d1);
        //MUSTALIAS(b1,a1);
//      foo(a,b,c,d);
//      foo(b,a,c,d);
}
```

Command to run :` gcc cs15.c`
`./a.out`

Reply:
It looks to me that `b1` points to `d1` and `d1` points to `a1`, i.e., MUSTALIAS(b1,&d1) and  MUSTALIAS(d1,&a1)

Reply:
![WhatsApp Image 2020-06-19 at 09 20 42](https://user-images.githubusercontent.com/43287592/85095171-4a0eed80-b20e-11ea-87ce-3c6d135ca01c.jpeg)
On paper I am getting this `b1->d` and `a1->d`.
Also the output also supports what I see on paper.
In case there is something wrong, please let me know.

Reply:
`b1 -> d1` but not `d`

Reply:
oops yes!
instead of d its **d1** in the function foo in the above _given picture_